### ISSUE TITLE
MCMC is slow/inefficient

### ISSUE THREAD

**Issue by yuvi-m** [MUSTANG-SZ/WITCH#218]
The MCMC sampler is much slower than expected.

**Comment by yuvi-m** [comment]
I dug into the MCMC slowness. The issue appears to be that `_leap`, `_sample`, and `_update` are defined inside `hmc` and decorated with `@jax.jit`. JAX caches compiled code keyed on the function object, and because these are new function objects on every `hmc` call, every call is a cache miss and recompiles from scratch. Because `_get_step_size` calls `hmc` up to `max_tries` times during tuning (plus main sampling call), we pay the full compilation cost multiple times instead of once.

Reproduction of the mechanism:
```python 
def bad_caller(x): 
  @jax.jit 
  def inner(v): return v*2 + 1   # redefined every call 
  return inner(x) 

@jax.jit
def good_inner(v): return v*2 + 1  # defined once 
def good_caller(x): return good_inner(x) 
```

With jax_log_compiles on: bad_caller compiles 4x over 4 calls, good_caller compiles 1x. This matches the hmc structure.

Proposed fix:
move `_leap`, `_sample`, and `_update` to module scope so they are defined once, passing the currently closed-over values (`log_prob`, `log_prob_grad`, `num_leaps`, `comm`) as explicit arguments, with the static ones handled via `static_argnums` / `partial`.

**Comment by yuvi-m** [comment]
@skhrg Want to confirm the proposed fix approach before I implement it?

**Comment by skhrg** [comment]
Yup this was my suspicion as well (see #163), we should make the same changes in the LM fitter.

**Comment by yuvi-m** [comment]
Okay, sounds good

**Comment by yuvi-m** [comment]
My plan is to hoist the private jitted functions (`_cond_func` and `_body_func` in `run_lmfit`, and `_leap`/`_sample`/`_update` in `hmc`) out to module scope so they're defined once instead of being recreated on every call. The values they currently close over (`tf`, `maxiter`, `chitol` for LM; `log_prob`, `num_leaps`, `comm` etc for HMC) would get passed in explicitly, with the static ones handled via `static_argnums`/`partial` so the `while_loop` /`fori_loop` interfaces still work. 

One caveat: I can't fully test this locally because my pipeline dies before fitting even starts. I hit the `run_pcg_wprior() got an unexpected keyword argument 'mask'` minkasi error during mapmaking, which looks like an environment/version thing unrelated to this. Since the whole point here is verifying that compile counts actually drip, I'd want to confirm this on the Penn machine that you told me about when describing the GPU project. I'll get started and open it as a draft so you can look at the approach, but I'll hold off on calling it done until I can profile it for real.

**Comment by Sulla2012** [comment]
That sounds reasonable to me. @skhrg was there a reason the functions were private initially? As for the `mask` issue you probably have to update WITCH; the mask keyword was introduced in PR #181 . Just try updating your WITCH installation. And yeah we can get you setup on PSD. Send @skhrg your preferred user and an ssh public key.

**Comment by skhrg** [comment]
No strong reason. Defining them privately is closer to the standard python practices for optimizer routines like this but as we have learned many time pythonic != jax-friendly...

**Comment by yuvi-m** [comment]
I got MCMC profiled locally (had to fix a few env issues and also a bug I'll put below). The results counter the recompilation hypothesis. 

I ran RXJ1347_a10.yaml + an mcmc block, num_steps=5, num_leaps=2, max_tries=3, on a reduced 1' grid at 2" pixels:

```
Trying step size 1e-08
Compiling MC sample function. This can take a few minutes!
Compiled MC sample function in 143.73823642730713 s
Trying step size 5e-08
Using cached sample function
Trying step size 1.5e-07
Using cached sample function
Using cached sample function
Using cached sample function
Took 1053.718380689621 s to run mcmc
```

hmc got called 5 times and compiled once; `_cache["hmc_sample"]`  is working. So, compilation is 144s of a 1054s run, meaning the other time is spent elsewhere. I don't think that hoisting `_leap`/`_sample`/`_update` to module scope would help here.

I believe that this points towards per-step sampling cost as opposed to compilation. Each leapfrog step calls log_prob_grad, which runs joint_objective across all TODs, and that happens num_leaps*2 times per step plus the MH correction. That might be something to look at.

Just as a note, I only ran max_tries:3, so only 5 hmc calls with small step counts. It may be possible that a different shape triggers a recompile I didn't hit.

Issue:
I think that `run_mcmc` is broken on main. `_log_prob`/`_log_prob_grad` (fitting.py, 470, 488) and the final updates (631, 637) call `MetaModel.update()` with 3 pos args, but it requires 4. It looks like the LM call sites got updated and the MCMC ones were missed. It throws TypeError on the first log-prob eval. I patched it locally by passing metamodel.cov to get the profiling done; I'll push that change through a separate PR, as it is unrelated.

**Comment by yuvi-m** [comment]
I have to walk back the timing numbers I posted. I did another run with the profiler on and got this:

```
1/1 [00:00<00:00,  2.55it/s]        -> 0.4 s/step
3/3 [04:21<00:00, 87.16s/it]        -> 87 s/step
5/5 [2:21:34<00:00, 1698.86s/it]    -> 1699 s/step
5/5 [02:48<00:00, 33.77s/it]        -> 34 s/step
5/5 [02:47<00:00, 33.45s/it]        -> 33 s/step
```

It was the same cached function, the same number of steps, the same data, but there was so much variation. The run also ended with signal 9 (killed), which I believe was a result of my laptop OOMing. So, the 1054s figure I said earlier is measuring the memory pressure on my computer, not WITCH. I also tried vmapping the per-parameter apply_noise loop in chisq_objective and saw a 0.6% difference, but between the two thrashing runs that doesn't really mean anything either.

So, I think that I can still do the structural analysis locally, but I can't actually make any performance numbers on this computer, which is the whole point of the issue. 

Could I get PSD access? I sent my username and ssh key over slack to Saianeesh a few days ago. Happy to keep poking at structure stuff in the meantime.

Also, separate from this, but run_mcmc is broken on main (the MetaModel.update() thing); I made a PR for that, and it's independent of the perf work (#220).

**Comment by Sulla2012** [comment]
Saianeesh should get you PSD access today. I think it'd be a good idea to do something even simpler than RXJ1347 to ensure that hardware capabilities are not seriously impacting the performance. I would do just a gaussian on a small grid. I would try to get consistent profiling results from that.

**Comment by yuvi-m** [comment]
That makes sense. I'm on PSD now so hopefully memory pressure won't be a factor anymore. I'll start with a single gaussian on a small grid like you suggested and make sure I can get repeatable timings for each step before touching anything else. 

Also, a couple of quick questions:

1. is there a shared WITCH environment or module on PSD i should use, or should I just set up my own venv and pip install?
2. Where does the MUSTANG-2 data live on this machine; WITCH_DATROOT isn't set and I don't see a WITCH_Data anywhere obvious. Want to make sure I point at the right shared data rather than copying anything.

**Comment by Sulla2012** [comment]
1. we should probably set one up but currently there isn't one.
2. Yeah, it lives at `/mnt/welch/MUSTANG/M2-TODs/`

**Comment by yuvi-m** [comment]
I got setup on PSD and have repeatable numbers now. I started with a single gaussian on a 1' grid at 2" pixels, no mapmaking, max_tries=0 so there's no tuning loop. 

I got two identical runs from this:

```
run1: compile 11.28s, 2.76 s/it, total 70.97s
run 2: compile 11.25s, 2.76 s/it, total 71.05s
```

Then, I varied num_leaps to see what scales and got:

```
num_leaps = 5: 2.76 s/it
num_leaps = 10: 4.86 s/it
```

To check where that goes, I timed joint_objective directly on the same loaded metamodel:

```
loglike only (MH correction): 159ms
grad only (leapfrog kick): 286 ms
all three (LM fit): 394 ms
```

Each leap does 2 grad calls and _sample does 2 loglike calls for the MH correction, so predicted per step is 10(286) + 2(159)=3.18s at num_leaps=5 and 20(286) + 2(159)=6.04s at 10. It measured 2.76 and 4.86, o slightly cheaper than naive but the same shape.

I think that the runtime is almost fully accounted for by joint_objective calls. These findings, combined with the earlier finding that hmc only compiles once; I don't know if there is a code level fix. On a real run at 5000 steps and num_leaps=10, its about 8 hours, which is almost all gradient eval.

The 286ms is CPU matmuls and DCTs in apply_noise, so switching to GPU seems like it will make the process a lot quicker. 

Note: the GPU on PSD is in an error state right now (nvidia-smi shows ERR! and culnit fails with CUDA_ERROR_NO_DEVICE) so all of the above testing is done on CPU.

**Comment by Sulla2012** [comment]
Ok in that case @skhrg can comment but I think the most important thing is to get GPU fitting working again. I think this was related to the issue where jax was taking things on and off memory too frequently but I can't find that Issue...

**Comment by yuvi-m** [comment]
Okay, sounds good; I can shift to getting GPU fitting working rather than chasing the CPU path. The PSD GPU is in an error state right now, though (nvidia smi shows ERR! on temp/power/util, and cuInit fails with CUDA_ERROR_NO_DEVICE), so I think it will need a reset before I can test on it. 

Saianeesh are you able to reset the GPU? If not, is there someone I should contact?

I also found another place where we are losing time: the gradient path computes and noise-filters all parameters and only slices down to the free ones at the end (6/15 free for RXJ1347), so fixed parameters get fully projected/filtered and then discarded. Isolating the noise-filter loop drops 2.2x going from 15 params to 6 params. It is probably worth masking down eventually since it applies on GPU too, but it's obviously secondary to getting GPU fitting back.

**Comment by skhrg** [comment]
GPU fitting works for me on PSD as of a few weeks ago.. though I see a ton of transfers back to host which are slowing things down. I'll take a look to see why cuda is mad.

Last I checked 1 proc on GPU was roughly as fast as 8 CPU procs even with the memory transfers.
@yuvi-m can you turn on the GPU memory transfer logging on your local copy and start to see where we could reduce them? First I think if you can save the logs and raise and issue with a list of the common transfers so we can start to take a look. I think a few well placed `put`s should fix things.

In parallel I will setup MPI on PSD built with CUDA support so we can start testing MPI on the GPU.

> I also found another place where we are losing time: the gradient path computes and noise-filters all parameters and only slices down to the free ones at the end (6/15 free for RXJ1347), so fixed parameters get fully projected/filtered and then discarded. Isolating the noise-filter loop drops 2.2x going from 15 params to 6 params.

Ah nice catch. I think the way to do this is to change this loop https://github.com/MUSTANG-SZ/WITCH/blob/1a3a6eb21f581c1522168cf770f69d5ef909ccfd/witch/objective.py#L175-L178 to only loop over the parameters we are fitting.

**Comment by yuvi-m** [comment]
I'm happy to take on the transfer logging. One block though, I can't actually run on GPU yet. My laptop can't handle the full pipeline, and the GPU on PSD is still in an error state (same as before), so JAX falls back to CPU. It has been like that since I first got access and tried, so I think the card needs a reset.

Once I have a working GPU I'll wrap a fit in jax.transfer_guard('log'), save the output, and raise an issue with the common transfers so we can see where the puts should go.

**Comment by yuvi-m** [comment]
I tried the projection half (slicing ip_grad to the fit rows in model_grad_proj before _project_vectorized, then scattering to par_map[free]). It runs, but it corrupts the gradient. The result differs from the unmasked gradient by around 6 on the fit params, so something in _project_vectorized or the metadata apply _grad/apply_grad_proj steps doens't tolerate row-slicing the way I assumed. I reverted the changes I made.

So for now, its just the filter-loop masking, which I've verified is correct (0 diff vs unmasked on the fit params, fixed params stay 0) and gives the 1.14x. The projection masking looks like it needs someone who knows what those projection/metadata transforms actually do the parameter axis; happy to pair on it if you want, but I didn't want to push a silent gradient bug into that path.

**Comment by yuvi-m** [comment]
Sorry, meant to post this before the comment above for context. I implemented the loop masking you pointed at and verified it's correct (0 diff vs unmasked on the fit params). On its own it gives around 1.14x on grad eval (1811 -> 1563 ms for RXJ1347_a10). It's smaller than the 2.2x from the isolated noise-loop test because model_grad_proj still projects all params, so the filter loop was only part of the gradient cost. That's what led me to try masking the projection too, which is the comment above.
